Sources/ProcedureKit/ProcedureFuture.swift

```diff
@@ -0,0 +1,334 @@
+
+ A simple example is:
+ 
+ ```swift
+ func myAsyncFunction() -> ProcedureFuture {
+    let promise = ProcedurePromise()
+    // dispatch an asynchronous task
+    DispatchQueue.global().async {
+        // do something
+        // then complete the promise when done
+        promise.complete()
+    }
+    return promise.future
+ }
+ 
+ myAsyncFunction().then(on: DispatchQueue.global()) {
+    // execute this block after myAsyncFunction's promise completes, on DispatchQueue.global()
+ }
+ ```
+ 
+ A ProcedureFuture/ProcedurePromise does not return a value - it merely allows you to execute 
+ a block when a promise is complete.
+ 
+ To return a future value, use ProcedureResultPromise (and ProcedureResultFuture).
+*/
+public class ProcedureFuture {
+    internal let group: DispatchGroup
+
+    internal init(group: DispatchGroup = DispatchGroup()) {
+        self.group = group
+        group.enter()
+    }
+
+    /// Schedules a block for execution after the ProcedureFuture completes.
+    ///
+    /// - Parameters:
+    ///   - on: a queue [DispatchQueue, EventQueue, Procedure (which schedules it on its internal event queue)]
+    ///   - block: The block to be scheduled after the ProcedureFuture completes.
+    public func then(on eventQueueProvider: QueueProvider, block: @escaping (Void) -> Void) {
+
+        let eventQueue = eventQueueProvider.providedQueue
+        eventQueue.dispatchNotify(withGroup: group, block: {
+            block()
+        })
+    }
+}
+
+// Used by ProcedurePromise
+fileprivate extension ProcedureFuture {
+
+    fileprivate func complete() {
+        group.leave()
+    }
+}
+
+// Used by ProcedureFutureGroup
+fileprivate extension ProcedureFuture {
+    // If the future is immediately available, execute the block synchronously on the current thread
+    // otherwise, queue a future dispatch onto the QueueProvider's queue.
+    @discardableResult
+    func thenOnSelfOrLater(on eventQueueProvider: QueueProvider, block: @escaping (Void) -> Void) -> ProcedureFuture {
+        let promise = ProcedurePromise()
+        let eventQueue = eventQueueProvider.providedQueue
+        if group.wait(timeout: .now()) == DispatchTimeoutResult.success {
+            // future is immediately available, execute the block synchronously on the current thread
+            block()
+            promise.complete()
+        }
+        else {
+            // future is not immediately available, so queue an asynchrous dispatch when it is on the
+            // provided queue
+            eventQueue.dispatchNotify(withGroup: group, block: {
+                block()
+                promise.complete()
+            })
+        }
+        return promise.future
+    }
+}
+
+/// Easily wait on multiple ProcedureFutures.
+/// Provides a future that is signaled once all ProcedureFutures in an array are signaled.
+///
+/// ```swift
+/// let futures: [ProcedureFuture] = ... // multiple ProcedureFutures
```

Comment:
Perhaps an extension on `Collection` or some such may make this even more convenient?